src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -29309,30 +29325,275 @@ class C<T>
             comp.VerifyDiagnostics();
         }
 
+        [Theory]
+        [InlineData("T")]
+        [InlineData("TNotNull")]
+        [WorkItem(39922, "https://github.com/dotnet/roslyn/issues/39922")]
+        public void EnforcedInMethodBody_MaybeNullWhen(string type)
+        {
+            var source = @"
+#nullable enable
+using System.Diagnostics.CodeAnalysis;
+
+class C<T, TClass, TNotNull>
+    where TClass : class
+    where TNotNull : notnull
+{
+    static void GetValue(TYPE x, [MaybeNull] out TYPE y)
+    {
+        y = x;
+    }
+
+    static bool TryGetValue(TYPE x, [MaybeNullWhen(true)] out TYPE y)
+    {
+        y = x;
+        return y == null;
+    }
+
+    static bool TryGetValue2(TYPE x, [MaybeNullWhen(true)] out TYPE y)
+    {
+        y = x;
+        return y != null;
+    }
+
+    static bool TryGetValue3(TYPE x, [MaybeNullWhen(false)] out TYPE y)
+    {
+        y = x;
+        return y == null;
+    }
+
+    static bool TryGetValue4(TYPE x, [MaybeNullWhen(false)] out TYPE y)
+    {
+        y = x;
+        return y != null;
+    }
+}
+";
+            var comp = CreateNullableCompilation(new[] { MaybeNullAttributeDefinition, MaybeNullWhenAttributeDefinition, source.Replace("TYPE", type) });
+            comp.VerifyDiagnostics();
+        }
+
         [Fact, WorkItem(39922, "https://github.com/dotnet/roslyn/issues/39922")]
-        public void EnforcedInMethodBody_Misc()
+        public void EnforcedInMethodBody_MaybeNullWhen_ReferenceType()
+        {
+            var source = @"
+#nullable enable
+using System.Diagnostics.CodeAnalysis;
+
+class C<TClass>
+    where TClass : class
+{
+    static bool TryGetValue(TClass x, [MaybeNullWhen(true)] out TClass y)
+    {
+        y = x;
+        return y == null;
+    }
+
+    static bool TryGetValue2(TClass x, [MaybeNullWhen(true)] out TClass y)
+    {
+        y = x;
+        return y != null; // 1
+    }
+
+    static bool TryGetValue3(TClass x, [MaybeNullWhen(false)] out TClass y)
+    {
+        y = x;
+        return y == null; // 2
+    }
+
+    static bool TryGetValue4(TClass x, [MaybeNullWhen(false)] out TClass y)
+    {
+        y = x;
+        return y != null;
+    }
+}
+";
+            // Note: this is a situation where it is annoying that we're learning from pure null tests
```

Comment:
I'm not sure I agree with this note. Each of the warnings is code that is at best misleading, and a likely source of annotation errors that could cascade once the attribute checks are fully implemented.

Reply:
Chatted offline with @jcouv and it's feeling like these warnings are good.

---
In reply to: [369130807](https://github.com/dotnet/roslyn/pull/41098#discussion_r369130807) [](ancestors = 369130807)

Reply:
➡️ removed comment